### PR TITLE
DDCE-4292: implement fallback mechanism for configurable number of months

### PR DESCRIPTION
**Pull request by rreddy80**
fallback implementation

### CODE REVIEW

File: app/uk/gov/hmrc/currencyconversion/repositories/ConversionRatePeriodJson.scala [hunk 1]
```diff
@@ -24,19 +24,33 @@ import uk.gov.hmrc.currencyconversion.utils.MongoIdHelper.currentFileName
...
 
-  def getExchangeRateFileName(date: LocalDate): Future[String] = {
-    val targetFileName = currentFileName(date)
+  private def getExchangeRateFileName(date: LocalDate): Future[String] = {
+    var fileName = "empty"
+    var fbMonth  = 0
+    while (fbMonth < fallbackMonths && fileName.equals("empty")) {
+      fileName = Await.result(checkFileExists(date.minusMonths(fbMonth)), Duration(maxWaitTime, SECONDS))
+      fbMonth += 1
+    }
```
Comment (lmugnai) [inline app/uk/gov/hmrc/currencyconversion/repositories/ConversionRatePeriodJson.scala:44]: Can this be refactored? try and stay away from using mutable variables

**Review by lmugnai — DISMISSED** [review]

File: test/uk/gov/hmrc/currencyconversion/controllers/ExchangeRateControllerSpec.scala [hunk 2]
```diff
@@ -86,10 +93,49 @@ class ExchangeRateControllerSpec
...
+    }
+    "return 200 and the correct json with scaling 2 decimal at least" in {
+      val result: Result = await(route(app, FakeRequest("GET", "/currency-conversion/rates/2019-12-10?cc=INR")).get)
+
+      result.header.status shouldBe OK
+
+      result.header.headers.get("Warning") shouldBe None
+
+      contentAsJson(Future.successful(result)) shouldBe Json.arr(
+        Json.obj("startDate" -> "2019-12-01", "endDate" -> "2019-12-31", "currencyCode" -> "INR", "rate" -> "1.00")
```
Comment (EdCope) [inline test/uk/gov/hmrc/currencyconversion/controllers/ExchangeRateControllerSpec.scala:118]: These tests are now duplicated I don't think you need these

**Review by EdCope — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/currencyconversion/repositories/ConversionRatePeriodJson.scala (hunk 1 above)
Comment (rreddy80) [reply app/uk/gov/hmrc/currencyconversion/repositories/ConversionRatePeriodJson.scala]: Sorry forgot to reply to this. Done, Lorenzo

File: app/uk/gov/hmrc/currencyconversion/services/ExchangeRateService.scala [hunk 3]
```diff
@@ -55,38 +55,25 @@ class ExchangeRateService @Inject() (exchangeRateRepository: ConversionRatePerio
...
     exchangeRateRepository.getCurrencyPeriod(date).flatMap {
-      case Some(value) => Future.successful(Some(value))
+      case Some(value) =>
+        logger.info(s"getCurrencyPeriod has returned a valid file....$value")
+        Future.successful(Some(value))
       case None        =>
-        val fallbackDate = date.minusMonths(1).`with`(lastDayOfMonth()) //lastDay of previous month
-        exchangeRateRepository.getCurrencyPeriod(fallbackDate)
+        logger.error(
+          s"XRS_FILE_NOT_AVAILABLE_ERROR [ConversionRatePeriodJson] No Exchange rate file is found with in the fallback period."
```
[the remark this answers is not in the document]
Comment (rreddy80) [reply app/uk/gov/hmrc/currencyconversion/services/ExchangeRateService.scala:74]: So this error is logged by ExchangeRateServie class only. Its not logged by any other class. why do we want to specify ConversionRatePeriodJson here? I have noticed some code where we unnecessarily specify the class name followed by a method name throwing error but in this case I felt its unnecessary as its all in the same class

[the remark this answers is not in the document]
Comment (EdCope) [reply app/uk/gov/hmrc/currencyconversion/services/ExchangeRateService.scala:74]: Sorry you're right think i copied from below !
I meant 
[ExchangeRateService] [getCurrencies] XRS_FILE_NOT_AVAILABLE_ERROR - No Exchange rate file is found within the fallback period.

File: test/uk/gov/hmrc/currencyconversion/connectors/HODConnectorSpec.scala [hunk 4]
```diff
@@ -69,6 +69,15 @@ class HODConnectorSpec
       connector.submit().futureValue.status mustBe OK
     }
 
+    "must fall back to a 503 error when downstream throws a bad request" in {
```
[the remark this answers is not in the document]
Comment (rreddy80) [reply test/uk/gov/hmrc/currencyconversion/connectors/HODConnectorSpec.scala:72]: this is to make sure all we return is a 503 irrespective of what the actual status is. There's a handler that handles bad requests (400 series). There's a dead code somewhere in the code that handles other errors, that needs to be removed as well in the future.

File: app/uk/gov/hmrc/currencyconversion/services/ExchangeRateService.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mywyau) [reply app/uk/gov/hmrc/currencyconversion/services/ExchangeRateService.scala:74]: Hey we should probably just re write these logs to something like:

logger.error(s"[ExchangeRateService][getRates] XRS_FILE_NOT_AVAILABLE_ERROR  Unable to get ConversionRatePeriod - Exchange rate file is not available") 

logger.error(s"[ExchangeRateService][getCurrencies] XRS_FILE_NOT_AVAILABLE_ERROR  Unable to get CurrencyPeriod - Exchange rate file is not available")

Tells us the class and method and can easily search which branch of code is being executed when we see the logs in kibana. But as for formatting prefer the "[classname][method] your log message goes here"  

if this comment clears up anything

File: app/uk/gov/hmrc/currencyconversion/repositories/ConversionRatePeriodJson.scala [hunk 5]
```diff
@@ -24,19 +24,38 @@ import uk.gov.hmrc.currencyconversion.utils.MongoIdHelper.currentFileName
...
+  private val fallbackMonths    = config.get[Int]("fallback.months")
+  private val maxWaitTime       = 5
+  private val emptyFile: String = "empty"
 
-  def getExchangeRateFileName(date: LocalDate): Future[String] = {
-    val targetFileName = currentFileName(date)
+  private def getExchangeRateFileName(date: LocalDate): Future[String] = {
+    @tailrec
+    def loop(givenDate: LocalDate): Future[String] = {
+      val returnedFile = Await.result(checkFileExists(givenDate), Duration(maxWaitTime, SECONDS))
```
Comment (mywyau) [inline app/uk/gov/hmrc/currencyconversion/repositories/ConversionRatePeriodJson.scala:43]: Not sure about blocking the Future and waiting for it. Does anyone know if we should be doing this? 

The recursion I think does not need the loop helper method as it's already tail recursive.

Screenshot 2023-06-23 at 12 26 20

File: test/uk/gov/hmrc/currencyconversion/controllers/ExchangeRateControllerUnitSpec.scala [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+class ExchangeRateControllerUnitSpec
+    extends AnyWordSpecLike
+    with GuiceOneAppPerSuite
+    with MockitoSugar
+    with BeforeAndAfterEach {
+
+  implicit val ec: scala.concurrent.ExecutionContext    = scala.concurrent.ExecutionContext.global
+  private lazy val exchangeRateRepository               = mock[ExchangeRateRepository]
+  val exchangeRateNew: Future[List[ExchangeRateResult]] = Future.successful(
```
[the remark this answers is not in the document]
Comment (rreddy80) [reply test/uk/gov/hmrc/currencyconversion/controllers/ExchangeRateControllerUnitSpec.scala]: deleted this file completely.

File: test/uk/gov/hmrc/currencyconversion/controllers/ExchangeRateControllerUnitSpec.scala [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      |                "validFrom" : "2019-09-01",
+      |                "validTo" : "2019-09-30",
+      |                "currencyCode" : "INR",
+      |                "exchangeRate" : 1,
+      |                "currencyName" : "India"
+      |            }
+      |        ]
+      |}""".stripMargin
+
+  "Getting rates for a date with no file received in the last 3 months and 1 valid currency" must {
```
[the remark this answers is not in the document]
Comment (rreddy80) [reply test/uk/gov/hmrc/currencyconversion/controllers/ExchangeRateControllerUnitSpec.scala]: this is no more used.

**Review by EdCope — APPROVED** [review]

**Pull request merged**